torch/lib/THCS/generic/THCSTensorMath.cu

```diff
@@ -24,38 +41,425 @@ void THCTensor_(spaddcdiv)(THCState *state, THCTensor *r_, THCTensor *t, real va
+    THCTensor *r_values_ = THCSTensor_(values)(state, r_);
+    THCTensor_(mul)(state, r_values_, r_values_, value);
+    THCTensor_(free)(state, r_values_);
+  } else {
+    THCSTensor_(resizeAs)(state, r_, t);
+
+    THCIndexTensor *r_indices_ = THCSTensor_(indices)(state, r_);
+    THCTensor *r_values_ = THCSTensor_(values)(state, r_);
+    THCIndexTensor *t_indices_ = THCSTensor_(indices)(state, t);
+    THCTensor *t_values_ = THCSTensor_(values)(state, t);
+
+    THCIndexTensor_(resizeAs)(state, r_indices_, t_indices_);
+    THCIndexTensor_(copy)(state, r_indices_, t_indices_);
+    THCTensor_(mul)(state, r_values_, t_values_, value);
+    r_->nnz = t->nnz;
+    r_->contiguous = t->contiguous;
+
+    THCIndexTensor_(free)(state, r_indices_);
+    THCTensor_(free)(state, r_values_);
+    THCIndexTensor_(free)(state, t_indices_);
+    THCTensor_(free)(state, t_values_);
+  }
 }
 
 void THCSTensor_(div)(THCState *state, THCSTensor *r_, THCSTensor *t, real value) {
-  THError("WARNING: Sparse Cuda Tensor op div is not implemented");
+  if (r_ == t) {
+    THCTensor *r_values_ = THCSTensor_(values)(state, r_);
+    THCTensor_(div)(state, r_values_, r_values_, value);
+    THCTensor_(free)(state, r_values_);
+  } else {
+    THCSTensor_(resizeAs)(state, r_, t);
+
+    THCIndexTensor *r_indices_ = THCSTensor_(indices)(state, r_);
+    THCTensor *r_values_ = THCSTensor_(values)(state, r_);
+    THCIndexTensor *t_indices_ = THCSTensor_(indices)(state, t);
+    THCTensor *t_values_ = THCSTensor_(values)(state, t);
+
+    THCIndexTensor_(resizeAs)(state, r_indices_, t_indices_);
+    THCIndexTensor_(copy)(state, r_indices_, t_indices_);
+    THCTensor_(div)(state, r_values_, t_values_, value);
+    r_->nnz = t->nnz;
+    r_->contiguous = t->contiguous;
+
+    THCIndexTensor_(free)(state, r_indices_);
+    THCTensor_(free)(state, r_values_);
+    THCIndexTensor_(free)(state, t_indices_);
+    THCTensor_(free)(state, t_values_);
+  }
 }
 
 void THCSTensor_(cadd)(THCState *state, THCSTensor *r_, THCSTensor *t, real value, THCSTensor *src) {
-  THError("WARNING: Sparse Cuda Tensor op cadd is not implemented");
+  THCAssertSameGPU(THCSTensor_(checkGPU)(state, 3, 3, r_, t, src));
+  if(!THCSTensor_(isSameSizeAs)(state, t, src)) {
+    THError("cadd operands have incompatible sizes or dimension types");
+  }
+  THCSTensor_(contiguous)(state, t);
+  THCSTensor_(contiguous)(state, src);
+
+  if (src->nnz == 0) {
+    THCSTensor_(copy)(state, r_, t);
+    return;
+  }
+  if (t->nnz == 0) {
+    THCSTensor_(mul)(state, r_, src, value);
+    return;
+  }
+
+  // We deliberately choose to simply concat the indices and values tensors
+  // rather than merging them. This removes the need to synchronously fetch nnz
+  // at the end of the operation, at the cost of having a non-contiguous result.
+  // This trade-off is preferable for the common use-case of gradient accumulation.
+  // TODO have two distinct functions? The other option is commented out below
```

Comment:
Are you sure that it's correct? What if there's a non-zero element in both these tensors? It's not only non-contiguous, it also has duplicate indices.